langstream-runtime/langstream-runtime-impl/src/main/java/ai/langstream/runtime/agent/GrpcAgentProcessor.java

```diff
@@ -0,0 +1,386 @@
+
+                if (record.key() != null) {
+                    recordBuilder.setKey(toGrpc(record.key()));
+                }
+
+                if (record.origin() != null) {
+                    recordBuilder.setOrigin(record.origin());
+                }
+
+                if (record.timestamp() != null) {
+                    recordBuilder.setTimestamp(record.timestamp());
+                }
+
+                if (record.headers() != null) {
+                    for (Header h : record.headers()) {
+                        ProcessorOuterClass.Header.Builder headerBuilder =
+                                recordBuilder.addHeadersBuilder().setName(h.key());
+                        if (h.value() != null) {
+                            headerBuilder.setValue(toGrpc(h.value()));
+                        }
+                    }
+                }
+                sourceRecords.put(rId, new RecordAndSink(record, recordSink));
+            } catch (Exception e) {
+                recordSink.emit(new SourceRecordAndResult(record, null, e));
+            }
+        }
+        if (recordsBuilder.getRecordCount() > 0) {
+            request.onNext(ProcessorRequest.newBuilder().setRecords(recordsBuilder).build());
+        }
+    }
+
+    @Override
+    public synchronized void close() throws Exception {
+        if (request != null) {
+            request.onCompleted();
+        }
+        if (channel != null) {
+            channel.shutdown();
+        }
+    }
+
+    private Object fromGrpc(Value value) throws IOException {
+        if (value == null) {
+            return null;
+        }
+        if (value.hasStringValue()) {
+            return value.getStringValue();
+        } else if (value.hasBytesValue()) {
+            return value.getBytesValue().toByteArray();
+        } else if (value.hasBooleanValue()) {
+            return value.getBooleanValue();
+        } else if (value.hasByteValue()) {
+            return (byte) value.getByteValue();
+        } else if (value.hasShortValue()) {
+            return (short) value.getShortValue();
+        } else if (value.hasIntValue()) {
+            return value.getIntValue();
+        } else if (value.hasLongValue()) {
+            return value.getLongValue();
+        } else if (value.hasFloatValue()) {
+            return value.getFloatValue();
+        } else if (value.hasDoubleValue()) {
+            return value.getDoubleValue();
+        } else if (value.hasJsonValue()) {
+            return value.getJsonValue();
+        } else if (value.hasAvroValue()) {
+            Object serverSchema = serverSchemas.get(value.getSchemaId());
+            if (serverSchema instanceof Schema schema) {
+                return deserializeGenericRecord(schema, value.getAvroValue().toByteArray());
+            } else {
+                // TODO: received unknown schema. error ?
+                // Do we send an error result or do we fail completely so that the agent is
+                // restarted ?
```

Comment:
What should we do here ?
Ignore ? Send error for the source record ? Fail the producer completely at the next process call ?

Reply:
fail completely

Reply:
OK.
Also currently I restart the request when the server errors.
But maybe it would be better to also fail completely to have the pod completely restarted.

Reply:
done